### DISCUSSION TITLE
Light Control

### DISCUSSION THREAD

**Discussion by Exit2Studios** [tagyoureit/nodejs-poolController#518] [Nixie]
I have PAL 2000 lights being controlled standalone (I don't have a Pentair Controller).  Is there a way to manage the color cycling via njsPC and/or REM?  I'm a bit unclear the best way to implement the lights and am getting an error currently.

<img width="473" alt="image" src="https://user-images.githubusercontent.com/16908875/177648651-841d2c59-706a-44b9-8701-c06dd65de476.png">

I've searched but can't really find the best way to setup lighting, which I thought would be self explanatory.  Once again, I get confused with circuits vs features, etc.

**Comment by rstrouse** [comment]
@nate-marshall set this up with PAL lighting.  #470 

I believe he set his to use the Pentair settings.

**Reply by Exit2Studios** [reply]
Mine are different.  There is no controller for them.  They simply cycle through the colors when they're turned on/off repeatedly.

**Reply by rstrouse** [reply]
Do you have a manual for them?

**Reply by Exit2Studios** [reply]
http://www.cherryhillpool.com/PAL-Treo%20LAU2%20Instructions%2083111.pdf

**Reply by rstrouse** [reply]
These operate in a similar fashion to the Hayward lights.  Set the type to ColorLogic and lets see if we can get the light to respond.  The color map will not be correct but we can start here.  How many ridiculous incarnations can these pool lighting people come up with.  All of them are less than poor from IntelliBrite to ColorLogic to PoolTone. 

Don't forget to select a relay by selecting the connection then finding the relay device in the dropdown.
![image](https://user-images.githubusercontent.com/47839015/177661787-354afdf5-0dbf-4f9a-8ce1-3848b10eebaf.png)

**Reply by Exit2Studios** [reply]
I agree with your assessment.  Somehow in messing with lights, I've created a bigger issue/error:

<img width="546" alt="image" src="https://user-images.githubusercontent.com/16908875/178033843-4c246e52-aca7-41dd-b1ac-54e24fd4dbf3.png">


`TypeError: Cannot read property 'master' of null
    at LightGroupCircuit.initData (/home/pi/nodejs-poolController/controller/Equipment.ts:456:46)
    at new EqItem (/home/pi/nodejs-poolController/controller/Equipment.ts:453:18)
    at new LightGroupCircuit (/home/pi/nodejs-poolController/controller/Equipment.ts:1751:1)
    at LightGroupCircuitCollection.createItem (/home/pi/nodejs-poolController/controller/Equipment.ts:1730:62)
    at LightGroupCircuitCollection.getItemByIndex (/home/pi/nodejs-poolController/controller/Equipment.ts:543:49)
    at LightGroupState.getExtended (/home/pi/nodejs-poolController/controller/State.ts:1317:37)
    at LightGroupStateCollection.getExtended (/home/pi/nodejs-poolController/controller/State.ts:668:69)
    at State.getState (/home/pi/nodejs-poolController/controller/State.ts:137:51)
    at /home/pi/nodejs-poolController/web/services/state/State.ts:458:40
    at Layer.handle [as handle_request] (/home/pi/nodejs-poolController/node_modules/express/lib/router/layer.js:95:5)
    at next (/home/pi/nodejs-poolController/node_modules/express/lib/router/route.js:144:13)
    at Route.dispatch (/home/pi/nodejs-poolController/node_modules/express/lib/router/route.js:114:3)
    at Layer.handle [as handle_request] (/home/pi/nodejs-poolController/node_modules/express/lib/router/layer.js:95:5)
    at /home/pi/nodejs-poolController/node_modules/express/lib/router/index.js:284:15
    at param (/home/pi/nodejs-poolController/node_modules/express/lib/router/index.js:365:14)
    at param (/home/pi/nodejs-poolController/node_modules/express/lib/router/index.js:376:14)`

If I go to the light group section, it will not allow me to delete the circuits, and when I save it gives me a null error.

**Reply by Exit2Studios** [reply]
<img width="634" alt="image" src="https://user-images.githubusercontent.com/16908875/178035578-78bf36d4-6cd9-4ce9-abc0-07753a6aaa07.png">

**Reply by Exit2Studios** [reply]
That is what shows up in Light Groups and I can't delete anything and if I try to add a name, etc and save it, I get a null value error.

**Reply by rstrouse** [reply]
Try hitting the trash can on the missing unnamed circuit in the group.   You are likely not going to be able to use that feature since your lights do not contain the command sequences for SAm lighting.   This is an IntelliBrite only feature.  The only thing a light group can do for you is turn on and off a group of lights.  It will not be able to do any SAm sequencing.

**Reply by Exit2Studios** [reply]
Hitting the trash can doesn't do anything.  I've tried using different browsers, restarted pi and even did a git pull (even though on current version).  Trash can doesn't work.

Is there a way to manually delete it from a config file somewhere?  I searched but came up empty.

**Reply by tagyoureit** [reply]
First, make a backup of your data files either manually or through the hamburger menu.

You can stop the app, the go into both the `./data/poolConfig.json` and `./data/poolState.json`.  Look for the `lightGroups:` section and delete any invalid entries.  You should be left with `"lightGroups": []` if you have no other light groups.
![2022-07-08_15-08-42](https://user-images.githubusercontent.com/1646840/178076493-f9f97c4f-f7d6-4a18-81a0-f5e1a9a15b93.png)

**Reply by Exit2Studios** [reply]
Interestingly, both were blank and now the problem seems to be resolved.  Maybe the new pull fixed it?  At any rate, thank you!

**Comment by spifftek70** [comment]
I have a few questions on this topic:

My pool has an color LED light that offers White, Blue, Green, and Red.  However the "control" for the color is set by flipping the light switch on and off to cycle through the color options.  Will the light controls within njsPC allow me to pick the color without cycling through power?  

The top is of the pool (color LED) and the bottom box is for the spa which is an older single color warm white light.  I don't see a transformer anywhere, so I believe the first example in the readme should work fine. I'm not sure what type of light I have, but I've provided pix of the light and the junction box.  Can someone tell me what kind / brand this is?

![IMG_0021](https://user-images.githubusercontent.com/11727190/194602876-ca7dee57-f3ba-4323-b6c0-5d3be5af0bec.jpeg)
<img width="1321" alt="Screen Shot 2022-10-07 at 9 25 42 AM" src="https://user-images.githubusercontent.com/11727190/194602901-66f35aee-c20d-413b-94ed-2db58638c66c.png">
<img width="848" alt="Screen Shot 2022-10-07 at 9 25 13 AM" src="https://user-images.githubusercontent.com/11727190/194602913-404bcede-51fa-4f1f-ac76-fba43c3fc9f9.png">

**Reply by tagyoureit** [reply]
The LED looks to be an [Intellibrite 5G](https://www.pentair.com/en-us/products/residential/pool-spa-equipment/pool-lighting/intellibrite_5g_underwatercolor-changingledpoolandspalights.html).  It probably has the additional 7 show colors, as well.  You can set this up when you define the circuit and be able to choose the color palette from dashPanel.  There isn't a way to directly select the color.  No matter what controls the light (switch, EasyTouch, njsPC) it still has to be controlled by sending a series of on-off pulses to cycle through the colors.  But the good news is that we take care of that for you.  You simply need to select the color.

If you other light is just a white light you can set it up as a `light` circuit and it will work just like a relay.

With both circuits, if you uncheck the `show as features` then the lights will move to a dedicated lights section in the UI.

**Reply by rstrouse** [reply]
Is that closeup of the color light or the white light?  That looks like a Pentair Amerlite which is only white with the standard bulb.  The IntelliBrite uses a different diffuser lens.

However if you installed a PoolTone replacement bulb then it will do color.  Select PoolTone for the circuit type.

**Reply by tagyoureit** [reply]
Ah, good call on the diffuser!

**Comment by spifftek70** [comment]
That’s a close-up of the colored LED lightDeanOn Oct 7, 2022, at 11:58 AM, rstrouse ***@***.***> wrote:﻿
Is that closeup of the color light or the white light?  That looks like a Pentair Amerlite which is only white with the standard bulb.  The IntelliBrite uses a different diffuser lens.
However if you installed a PoolTone replacement bulb then it will do color.  Select PoolTone for the circuit type.

**Reply by rstrouse** [reply]
Does it have one of these bulbs installed in the niche?


https://www.florida-sunseeker.com/collections/upgrade-kit/products/pentair-16-color-led-upgrade-kit-for-amerlite-large-pool-size?variant=6448912269339

**Reply by spifftek70** [reply]
The bulb was installed already when I purchased the house. I have no idea what the brand is.

**Reply by rstrouse** [reply]
The simplest way to find out is to hook this up to a relay and set it up in REM.  Then defined a circuit in dashPanel that has a type of PoolTone.  When you click on the color palette it will show you the options.

![image](https://user-images.githubusercontent.com/47839015/194669287-ff7bb383-1b48-4185-bcea-28319c7213c7.png)

**Comment by Exit2Studios** [comment]
It would be great if you could define custom lights (pick the colors and order they cycle). Maybe this is already implemented. I haven’t been around here for a while.
________________________________

**Reply by rstrouse** [reply]
I agree but the pool industry is good at creating beautiful brochures regarding lighting but really crappy equipment.  This on-off-on protocol that they all seem to have adopted is flatline brain dead.  What needs happen is for someone to make a light engine that conforms to standard lighting protocols.  Given the fact that they charge $600 to $1k per fixture you would think that they could do something innovative.

When my IntelliBrites fail they will get DMX encoders in the can.

**Reply by Telemachus1977** [reply]
I have been saying something similar about my incandescents for 15 years now...  But they never die.

They just don't make them like they used to.
